File: treeadapter/src/main/java/com/krystianwsul/treeadapter/TreeNodeCollection.java
package com.krystianwsul.treeadapter;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import com.annimon.stream.Collectors;
import com.annimon.stream.Stream;
import com.annimon.stream.function.Consumer;
import com.annimon.stream.function.Function;

import junit.framework.Assert;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class TreeNodeCollection implements NodeContainer {
    @Nullable
    private List<TreeNode> mTreeNodes;

    @NonNull
    final TreeViewAdapter mTreeViewAdapter;

    public TreeNodeCollection(@NonNull TreeViewAdapter treeViewAdapter) {
        mTreeViewAdapter = treeViewAdapter;
    }

    public void setNodes(@NonNull List<TreeNode> rootTreeNodes) {
        if (mTreeNodes != null)
            throw new SetTreeNodesCalledTwiceException();

        mTreeNodes = new ArrayList<>(rootTreeNodes);

        Collections.sort(mTreeNodes);
    }

    @NonNull
    TreeNode getNode(int position) {
        if (mTreeNodes == null)
            throw new SetTreeNodesNotCalledException();

        Assert.assertTrue(position >= 0);
        Assert.assertTrue(position < displayedSize());

        for (TreeNode treeNode : mTreeNodes) {
            if (position < treeNode.displayedSize())
                return treeNode.getNode(position);

            position = position - treeNode.displayedSize();
        }

        throw new IndexOutOfBoundsException();
    }

    @Override
    public int getPosition(@NonNull TreeNode treeNode) {
        if (mTreeNodes == null)
            throw new SetTreeNodesNotCalledException();

        int offset = 0;
        for (TreeNode currTreeNode : mTreeNodes) {
            int position = currTreeNode.getPosition(treeNode);
            if (position >= 0)
                return offset + position;
            offset += currTreeNode.displayedSize();
        }

        return -1;
    }

    int getItemViewType(int position) {
        TreeNode treeNode = getNode(position);

        return treeNode.getItemViewType();
    }

    @Override
    public int displayedSize() {
        if (mTreeNodes == null)
            throw new SetTreeNodesNotCalledException();

        int displayedSize = 0;
        for (TreeNode treeNode : mTreeNodes)
            displayedSize += treeNode.displayedSize();
        return displayedSize;
    }

    @NonNull
    List<TreeNode> getSelectedNodes() {
        if (mTreeNodes == null)
            throw new SetTreeNodesNotCalledException();

        return Stream.of(mTreeNodes)
                .flatMap(new Function<TreeNode, Stream<TreeNode>>() {
                    @Override
                    public Stream<TreeNode> apply(TreeNode treeNode) {
                        return treeNode.getSelectedNodes();
                    }
                })
                .collect(Collectors.<TreeNode>toList());
    }

    void onCreateActionMode() {
        if (mTreeNodes == null)
            throw new SetTreeNodesNotCalledException();

        Stream.of(mTreeNodes)
                .forEach(new Consumer<TreeNode>() {
                    @Override
                    public void accept(TreeNode treeNode) {
                        treeNode.onCreateActionMode();
                    }
                });
    }

    void onDestroyActionMode() {
        if (mTreeNodes == null)
            throw new SetTreeNodesNotCalledException();

        Stream.of(mTreeNodes)
                .forEach(new Consumer<TreeNode>() {
                    @Override
                    public void accept(TreeNode treeNode) {
                        treeNode.onDestroyActionMode();
                    }
                });
    }

    void unselect() {
        if (mTreeNodes == null)
            throw new SetTreeNodesNotCalledException();

        Stream.of(mTreeNodes)
                .forEach(new Consumer<TreeNode>() {
                    @Override
                    public void accept(TreeNode treeNode) {
                        treeNode.unselect();
                    }
                });
    }

    @Override
    public void add(@NonNull TreeNode treeNode) {
        if (mTreeNodes == null)
            throw new SetTreeNodesNotCalledException();

        mTreeNodes.add(treeNode);

        Collections.sort(mTreeNodes);

        TreeViewAdapter treeViewAdapter = mTreeViewAdapter;

        int newPosition = getPosition(treeNode);
        Assert.assertTrue(newPosition >= 0);

        treeViewAdapter.notifyItemInserted(newPosition);

        if (newPosition > 0)
            treeViewAdapter.notifyItemChanged(newPosition - 1);
    }

    @Override
    public void remove(@NonNull TreeNode treeNode) {
        if (mTreeNodes == null)
            throw new SetTreeNodesNotCalledException();

        Assert.assertTrue(mTreeNodes.contains(treeNode));

        TreeViewAdapter treeViewAdapter = mTreeViewAdapter;

        int oldPosition = getPosition(treeNode);
        Assert.assertTrue(oldPosition >= 0);

        int displayedSize = treeNode.displayedSize();

        mTreeNodes.remove(treeNode);

        treeViewAdapter.notifyItemRangeRemoved(oldPosition, displayedSize);

        if (oldPosition > 0)
            treeViewAdapter.notifyItemChanged(oldPosition - 1);
    }

    @Override
    public boolean expanded() {
        return true;
    }

    @Override
    public void update() {

    }

    @Override
    public void updateRecursive() {

    }

    @NonNull
    @Override
    public List<TreeNode> getSelectedChildren() {
        return getSelectedNodes();
    }

    @NonNull
    @Override
    public TreeNodeCollection getTreeNodeCollection() {
        return this;
    }

    void selectAll() {
        if (mTreeNodes == null)
            throw new SetTreeNodesNotCalledException();

        Stream.of(mTreeNodes)
                .forEach(new Consumer<TreeNode>() {
                    @Override
                    public void accept(TreeNode treeNode) {
                        treeNode.selectAll();
                    }
                });
    }

    @Override
    public int getIndentation() {
        return 0;
    }

    @SuppressWarnings("WeakerAccess")
    public static class SetTreeNodesNotCalledException extends InitializationException {
        private SetTreeNodesNotCalledException() {
            super("TreeNodeCollection.setTreeNodes() has not been called.");
        }
    }

    @SuppressWarnings("WeakerAccess")
    public static class SetTreeNodesCalledTwiceException extends InitializationException {
        private SetTreeNodesCalledTwiceException() {
            super("TreeNodeCollection.setTreeNodes() has already been called.");
        }
    }
}

File: treeadapter/src/main/java/com/krystianwsul/treeadapter/TreeNodeCollection.kt
package com.krystianwsul.treeadapter

import com.jakewharton.rxrelay3.BehaviorRelay
import com.krystianwsul.common.utils.Ordinal
import io.reactivex.rxjava3.core.Observable
import kotlin.math.min

class TreeNodeCollection<T : TreeHolder>(val treeViewAdapter: TreeViewAdapter<T>) : NodeContainer<T> {

    private val treeNodesRelay = BehaviorRelay.create<List<TreeNode<T>>>()

    val selectedNodes
        get() = treeNodesRelay.value
            ?.flatMap { it.selectedNodes }
            ?: throw SetTreeNodesNotCalledException()

    val nodesObservable: Observable<List<TreeNode<T>>> = treeNodesRelay

    var nodes
        get() = treeNodesRelay.value ?: throw SetTreeNodesNotCalledException()
        set(rootTreeNodes) {
            check(treeViewAdapter.locker == null)

            if (treeNodesRelay.value != null) throw SetTreeNodesCalledTwiceException()

            treeNodesRelay.accept(rootTreeNodes.sorted())
        }

    override val wantsSeparators = false

    override fun getNode(position: Int, positionMode: PositionMode): TreeNode<T> {
        check(position >= 0)

        return positionMode.getRecursiveNodes(this)[position]
    }

    override fun getPosition(positionMode: PositionMode, matcher: (TreeNode<T>) -> Boolean): Int {
        val treeNodes = positionMode.getDirectChildNodes(this)

        var offset = 0
        for (currTreeNode in treeNodes) {
            val position = currTreeNode.getPosition(positionMode, matcher)
            if (position >= 0) return offset + position

            offset += positionMode.getRecursiveNodes(currTreeNode).size
        }

        return -1
    }

    fun getItemViewType(position: Int): Int {
        val treeNode = getNode(position)

        return treeNode.itemViewType
    }

    override val displayedNodes get() = nodes.flatMap { it.displayedNodes }

    override val displayedChildNodes get() = displayedNodes

    override val displayableNodes
        get() = treeNodesRelay.value
            ?.flatMap { it.displayableNodes }
            ?: throw SetTreeNodesNotCalledException()

    override val displayedDirectChildNodes get() = nodes.filter { it.visible() }

    fun unselect(x: TreeViewAdapter.Placeholder) = treeNodesRelay.value
        ?.forEach { it.unselect(x) }
        ?: throw SetTreeNodesNotCalledException()

    override fun add(treeNode: TreeNode<T>, placeholder: TreeViewAdapter.Placeholder) {
        val treeNodes = treeNodesRelay.value
            ?.toMutableList()
            ?: throw SetTreeNodesNotCalledException()

        treeNodes.add(treeNode)
        treeNodes.sort()

        check(treeViewAdapter.locker == null)

        treeNodesRelay.accept(treeNodes)
    }

    override fun remove(treeNode: TreeNode<T>, placeholder: TreeViewAdapter.Placeholder) {
        val treeNodes = treeNodesRelay.value
            ?.toMutableList()
            ?: throw SetTreeNodesNotCalledException()

        check(treeNodes.contains(treeNode))

        treeNodes.remove(treeNode)

        check(treeViewAdapter.locker == null)

        treeNodesRelay.accept(treeNodes)
    }

    override val isExpanded = true

    override val selectedChildren get() = selectedNodes

    override val treeNodeCollection = this

    fun selectAll(placeholder: TreeViewAdapter.Placeholder) = displayedNodes.forEach { it.select(placeholder) }

    override val indentation = 0

    override val showInheritableBottomSeparator = false

    fun moveItem(fromPosition: Int, toPosition: Int, placeholder: TreeViewAdapter.Placeholder) {
        val fromTreeNode = getNode(fromPosition, PositionMode.DISPLAYED)
        val toTreeNode = getNode(toPosition, PositionMode.DISPLAYED)

        val fromParent = fromTreeNode.parent
        val toParent = toTreeNode.parent

        if (fromParent == toParent) {
            fromParent.swapNodePositions(fromTreeNode, toTreeNode, placeholder)
        } else {
            fromParent.removeForSwap(fromTreeNode, placeholder)
            toParent.addForSwap(fromTreeNode, toTreeNode, placeholder)
        }
    }

    override fun swapNodePositions(
        fromTreeNode: TreeNode<T>,
        toTreeNode: TreeNode<T>,
        placeholder: TreeViewAdapter.Placeholder,
    ) {
        check(treeViewAdapter.locker == null)

        val treeNodes = treeNodesRelay.value
            ?.toMutableList()
            ?: throw SetTreeNodesNotCalledException()

        val fromPosition = treeNodes.indexOf(fromTreeNode)
        check(fromPosition >= 0)

        val toPosition = treeNodes.indexOf(toTreeNode)
        check(toPosition >= 0)

        treeNodes.apply {
            removeAt(fromPosition)
            add(toPosition, fromTreeNode)
        }

        treeNodesRelay.accept(treeNodes)
    }

    override fun removeForSwap(fromTreeNode: TreeNode<T>, placeholder: TreeViewAdapter.Placeholder) {
        check(treeViewAdapter.locker == null)

        val treeNodes = treeNodesRelay.value
            ?.toMutableList()
            ?: throw SetTreeNodesNotCalledException()

        val fromPosition = treeNodes.indexOf(fromTreeNode)
        check(fromPosition >= 0)

        treeNodes.removeAt(fromPosition)

        treeNodesRelay.accept(treeNodes)
    }

    override fun addForSwap(fromTreeNode: TreeNode<T>, toTreeNode: TreeNode<T>, placeholder: TreeViewAdapter.Placeholder) {
        check(treeViewAdapter.locker == null)

        val treeNodes = treeNodesRelay.value
            ?.toMutableList()
            ?: throw SetTreeNodesNotCalledException()

        val toPosition = treeNodes.indexOf(toTreeNode)
        check(toPosition >= 0)

        fromTreeNode.parent = this
        treeNodes.add(toPosition, fromTreeNode)

        treeNodesRelay.accept(treeNodes)
    }

    private fun getAdjacentNodes(currentTreeNode: TreeNode<*>): Pair<Sortable?, Sortable?> {
        val displayedNodesInParent = currentTreeNode.parent.displayedDirectChildNodes
        val currentPositionInParent = displayedNodesInParent.indexOf(currentTreeNode)

        val previousNode = currentPositionInParent.takeIf { it > 0 }
            ?.let { displayedNodesInParent[currentPositionInParent - 1] }
            ?.modelNode
            ?.let { it as? Sortable }
            ?.takeIf { it.sortable }

        val nextNode = currentPositionInParent.takeIf { it < displayedNodesInParent.size - 1 }
            ?.let { displayedNodesInParent[currentPositionInParent + 1] }
            ?.modelNode
            ?.let { it as? Sortable }
            ?.takeIf { it.sortable }

        val reversedOrdinal = listOfNotNull(previousNode, nextNode).map { it.reversedOrdinal }
            .distinct()
            .single()

        return if (reversedOrdinal)
            nextNode to previousNode
        else
            previousNode to nextNode
    }

    fun setNewItemPosition(position: Int) {
        val allDisplayedNodes = displayedNodes
        val adjustedPositionInAllNodes = min(allDisplayedNodes.size - 1, position) // padding
        val currentTreeNode = allDisplayedNodes[adjustedPositionInAllNodes]
        val currentNode = currentTreeNode.modelNode

        val (previousNode, nextNode) = getAdjacentNodes(currentTreeNode)

        check(previousNode != null || nextNode != null)

        val previousOrdinal: Ordinal
        val nextOrdinal: Ordinal
        if (previousNode != null) {
            previousOrdinal = previousNode.getOrdinal()

            val lastOrdinal = if (previousOrdinal.isInt())
                previousOrdinal + 1
            else
                previousOrdinal.ceiling()

            nextOrdinal = nextNode?.getOrdinal() ?: lastOrdinal
        } else {
            nextOrdinal = (nextNode as Sortable).getOrdinal()
            previousOrdinal = nextOrdinal - 1000
        }

        check(nextOrdinal > previousOrdinal)

        (currentNode as Sortable).setOrdinal((previousOrdinal + nextOrdinal) / 2)
    }

    fun selectNode(position: Int) = treeViewAdapter.updateDisplayedNodes { displayedNodes[position].select(it) }

    fun resetExpansion(onlyProgrammatic: Boolean, placeholder: TreeViewAdapter.Placeholder) {
        treeNodesRelay.value!!.forEach { it.resetExpansion(onlyProgrammatic, placeholder) }
    }

    fun expandMatching() = treeNodesRelay.value!!.forEach { it.expandMatching() }

    class SetTreeNodesNotCalledException : InitializationException("TreeNodeCollection.setTreeNodes() has not been called.")

    class SetTreeNodesCalledTwiceException :
        InitializationException("TreeNodeCollection.setTreeNodes() has already been called.")

    override val id = Id

    object Id
}
